src/memory/mmap_bytebuf.rs

```diff
@@ -5,181 +5,125 @@
-                    .ok_or_else(|| "mmap returned 0".to_string())?;
-                Ok(Self { ptr, len })
-            }
-        }
-    }
-
-    fn as_slice(&self) -> &[u8] {
-        unsafe {
-            // Safety Proof:
-            // - Aliasing guarantees of `self.ptr` are not violated since `self` is the only owner.
-            // - This pointer was allocated for `self.len` bytes and thus is a valid slice.
-            // - `self.len` doesn't change throughout the lifetime of `self`.
-            // - The value is returned valid for the duration of lifetime of `self`.
-            //   `self` cannot be destroyed while the returned slice is alive.
-            // - `self.ptr` is of `NonNull` type and thus `.as_ptr()` can never return NULL.
-            // - `self.len` cannot be larger than `isize::max_value()`.
-            slice::from_raw_parts(self.ptr.as_ptr(), self.len)
-        }
+        let allocation =
+            region::alloc(len, Protection::READ_WRITE).map_err(|error| error.to_string())?;
+        Ok(Self { allocation })
     }
 
-    fn as_slice_mut(&mut self) -> &mut [u8] {
-        unsafe {
-            // Safety Proof:
-            // - See the proof for `Self::as_slice`
-            // - Additionally, it is not possible to obtain two mutable references for `self.ptr`
-            slice::from_raw_parts_mut(self.ptr.as_ptr(), self.len)
-        }
+    /// Returns a shared slice over the bytes of the virtual memory allocation.
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        // # SAFETY
+        //
+        // The operation is safe since we assume that the virtual memory allocation
+        // has been successful and allocated exactly `self.allocation.len()` bytes.
+        // Therefore creating a slice with `self.len` elements is valid.
+        // Aliasing guarantees are not violated since `self` is the only owner
+        // of the underlying virtual memory allocation.
+        unsafe { slice::from_raw_parts(self.allocation.as_ptr(), self.allocation.len()) }
     }
-}
 
-impl Drop for Mmap {
-    fn drop(&mut self) {
-        let ret_val = unsafe {
-            // Safety proof:
-            // - `self.ptr` was allocated by a call to `mmap`.
-            // - `self.len` was saved at the same time and it doesn't change throughout the lifetime
-            //   of `self`.
-            libc::munmap(self.ptr.as_ptr() as *mut libc::c_void, self.len)
-        };
-
-        // There is no reason for `munmap` to fail to deallocate a private annonymous mapping
-        // allocated by `mmap`.
-        // However, for the cases when it actually fails prefer to fail, in order to not leak
-        // and exhaust the virtual memory.
-        assert_eq!(ret_val, 0, "munmap failed");
+    /// Returns an exclusive slice over the bytes of the virtual memory allocation.
+    #[inline]
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        // # SAFETY
+        //
+        // See safety proof of `Mmap::as_slice`.
```

Comment:
```suggestion
        // See safety proof of `VirtualMemory::as_slice`.
```

or

```suggestion
        // See safety proof of `Self::as_slice`.
```

?

Reply:
good catch. but none of the above since this is a pitfall for renaming the type.